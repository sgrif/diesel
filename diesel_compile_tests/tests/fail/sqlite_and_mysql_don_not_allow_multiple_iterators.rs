extern crate diesel;

use diesel::prelude::*;
use diesel::sql_query;

fn main() {
    let conn = &mut SqliteConnection::establish("foo").unwrap();
    // For sqlite the returned iterator is coupled to
    // a statement, which is coupled to the connection itself
    // so we cannot have more than one iterator
    // for the same connection
    let row_iter1 = conn.load(&sql_query("bar")).unwrap();
    let row_iter2 = conn.load(&sql_query("bar")).unwrap();

    let _ = row_iter1.zip(row_iter2);

    let conn = &mut MysqlConnection::establish("foo").unwrap();
    // The same argument applies to mysql
    let row_iter1 = conn.load(&sql_query("bar")).unwrap();
    let row_iter2 = conn.load(&sql_query("bar")).unwrap();

    let _ = row_iter1.zip(row_iter2);

    let conn = &mut PgConnection::establish("foo").unwrap();
    // It works for PgConnection as the result is not related to the
    // connection in any way
    let row_iter1 = conn.load(&sql_query("bar")).unwrap();
    let row_iter2 = conn.load(&sql_query("bar")).unwrap();

    let _ = row_iter1.zip(row_iter2);

}
