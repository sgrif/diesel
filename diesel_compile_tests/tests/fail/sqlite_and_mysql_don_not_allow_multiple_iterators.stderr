error[E0499]: cannot borrow `*conn` as mutable more than once at a time
  --> $DIR/sqlite_and_mysql_don_not_allow_multiple_iterators.rs:13:21
   |
12 |     let row_iter1 = conn.load(&sql_query("bar")).unwrap();
   |                     ---- first mutable borrow occurs here
13 |     let row_iter2 = conn.load(&sql_query("bar")).unwrap();
   |                     ^^^^ second mutable borrow occurs here
14 |
15 |     let _ = row_iter1.zip(row_iter2);
   |             --------- first borrow later used here

error[E0499]: cannot borrow `*conn` as mutable more than once at a time
  --> $DIR/sqlite_and_mysql_don_not_allow_multiple_iterators.rs:20:21
   |
19 |     let row_iter1 = conn.load(&sql_query("bar")).unwrap();
   |                     ---- first mutable borrow occurs here
20 |     let row_iter2 = conn.load(&sql_query("bar")).unwrap();
   |                     ^^^^ second mutable borrow occurs here
21 |
22 |     let _ = row_iter1.zip(row_iter2);
   |             --------- first borrow later used here
